mesonbuild/interpreter.py

```diff
@@ -2688,6 +2684,12 @@ def add_languages(self, args, required):
                 mlog.log('Cross', cross_comp.get_display_language(), 'compiler:',
                          mlog.bold(' '.join(cross_comp.get_exelist())), version_string)
                 self.build.ensure_static_cross_linker(comp)
+
+        langs = self.coredata.compilers.keys()
+        if 'vala' in langs:
+            if 'c' not in langs:
+                raise InterpreterException('Compiling Vala requires C. Add C to your project languages and rerun Meson.')
```

Comment:
With this change, I think this won't work:
```meson
project('foo', 'c')
add_languages('vala')
```

Reply:
More corner case, but what about `add_languages('vala', 'c', required : false)` and the `c` compiler cannot be found, but the `vala` one is found?

Reply:
That example should work. If it does not, then it needs to be fixed before merging.

Reply:
In fact changing one of the existing Vala unit tests to do `add_languages` would be a good change in any case.

Reply:
ping

Reply:
`langs` here is the list of active languages, not the list supplied to `add_languages()`, so both cases presented here should work.  

I've extended a test to cover the first, the second is much harder to write a test for, as we'd need to make the C compiler unavailable somehow.